mne/beamformer/_ap.py

```diff
@@ -0,0 +1,765 @@
+                                           sub_g_proj)
+                                 ),
+                            sub_g_proj.transpose()])
+        perpend_spc = np.eye(act_spc.shape[0]) - act_spc
+        for dip in range(attr_dict['ndipoles']):
+            if force_no_rep and (dip in s_ap):
+                continue
+            sol_tuple = _solve_active_gain_eig(dip, gain,
+                                               data_cov, eig,
+                                               perpend_spc)
+            ap_val2[dip] = np.max([x.real for x in sol_tuple[0]])
+
+        s2_idx = np.argmax(ap_val2)
+        s_ap.append(s2_idx)
+        sol_tuple = _solve_active_gain_eig(s2_idx, gain,
+                                           data_cov, eig,
+                                           perpend_spc)
+        oris[src] = sol_tuple[1][:,
+                                 [np.argmax([x.real for x
+                                            in sol_tuple[0]])]
+                                 ][:, 0]
+        sub_g_proj = np.concatenate([sub_g_proj,
+                                     np.dot(sol_tuple[2],
+                                            oris[src])[:, np.newaxis]
+                                     ],
+                                    axis=1)
+    return s_ap, oris, sub_g_proj
+
+
+def _free_phase2(ap_temp_tuple, attr_dict,
+                 data_cov, gain, force_no_rep):
+    """Calculate phase 2 of free oriented AP.
+
+    altering the projection of current estimated dipoles
+
+    Parameters
+    ----------
+    ap_temp_tuple : tuple
+        See: _free_phase1b.
+    attr_dict : dict
+        See: _calculate_free_alternating_projections.
+    data_cov : array
+        Data Covariance.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+    force_no_rep : bool
+        Forces no repetition of estinated dipoles.
+
+    Returns
+    -------
+    s_ap_2 : list of int
+        List of dipole indices.
+    oris : array, shape (nsources, 3)
+        Orientations array of estimated sources (sorted by s_ap_2).
+    sub_g_proj : array
+        Sub space projected by estimated dipoles.
+
+    """
+    # ap_temp_tuple = (s_ap, oris, sub_g_proj)
+    from scipy.linalg import eig
+
+    s_ap_2, oris, sub_g_proj = copy(ap_temp_tuple)
+    print(' 2nd phase : ')
+    for itr in range(attr_dict['max_iter']):
+        print('iteration No. {}'.format(itr + 1))
+        s_ap_2_prev = copy(s_ap_2)
+        for src in range(attr_dict['nsources']):
+            # AP localization of src-th source
+            ap_val2 = np.zeros(attr_dict['ndipoles'])
+            a_tmp = copy(ap_temp_tuple[2])
+            a_tmp = np.delete(a_tmp, src, 1)
+            act_spc = multi_dot([a_tmp,
+                                 pinv(np.matmul(a_tmp.transpose(),
+                                                a_tmp)
+                                      ),
+                                 a_tmp.transpose()])
```

Comment:
here's that `multi_dot([a, pinv(a.T @ a), a.T])` pattern again. please simplify or use helper function

Reply:
created an helper function which replaced this pattern in 4 places, also - changed all multi_dot() usages to be called on one line.